src/test/scala/sangria/schema/ArgsSpec.scala

```diff
@@ -0,0 +1,248 @@
+        args.raw should be (Map.empty)
+        args.argsWithDefault should be (Set.empty)
+        args.optionalArgs should be (Set(OptionalArgumentName))
+        args.undefinedArgs should be (Set(OptionalArgumentName))
+        args.defaultInfo should be (TrieMap.empty)
+
+        args.arg(optionalArgument) should be (None)
+      }
+
+      "build with overriden default values" in {
+        val args = Args(List(defaultArgument), s"""{"$DefaultArgumentName": 9001}""".parseJson)
+        args.raw should be (Map(DefaultArgumentName -> Some(9001)))
+        args.argsWithDefault should be (Set(DefaultArgumentName))
+        args.optionalArgs should be (Set(DefaultArgumentName))
+        args.undefinedArgs should be (Set.empty)
+        args.defaultInfo should be (TrieMap.empty)
+
+        args.arg(defaultArgument) should be (9001)
+      }
+    }
+
+    "buildArgs with nested objects" should {
+      "build with nested arguments" in {
+        val json =
+          s"""
+             |{
+             |  "$NestedParentArgumentName": {
+             |    "$NonDefaultArgumentName": 1,
+             |    "$DefaultArgumentName": 2,
+             |    "$OptionalArgumentName": 3
+             |  }
+             |}
+           """.stripMargin.parseJson
+        val args = Args(List(nestedParentArgument), json)
+        val fields = args.arg(nestedParentArgument).asJsObject.fields
+
+        fields(NonDefaultArgumentName) should be (JsNumber(1))
+        fields(DefaultArgumentName) should be (JsNumber(2))
+        fields(OptionalArgumentName) should be (JsNumber(3))
+      }
+
+      "not build without required arguments" in {
+        val json =
+          s"""
+             |{
+             |  "$NestedParentArgumentName": {
+             |    "$DefaultArgumentName": 2,
+             |    "$OptionalArgumentName": 3
+             |  }
+             |}
+           """.stripMargin.parseJson
+
+        an [AttributeCoercionError] should be thrownBy Args(List(nestedParentArgument), json)
+      }
+
+      "build without default arguments" in {
+        val json =
+          s"""
+             |{
+             |  "$NestedParentArgumentName": {
+             |    "$NonDefaultArgumentName": 1,
+             |    "$OptionalArgumentName": 3
+             |  }
+             |}
+           """.stripMargin.parseJson
+
+        val args = Args(List(nestedParentArgument), json)
+        val fields = args.arg(nestedParentArgument).asJsObject.fields
+
+        fields(NonDefaultArgumentName) should be (JsNumber(1))
+        fields(DefaultArgumentName) should be (JsNumber(10))
+        fields(OptionalArgumentName) should be (JsNumber(3))
+      }
+
+      "build without optional arguments" in {
+        val json =
+          s"""
+             |{
+             |  "$NestedParentArgumentName": {
+             |    "$NonDefaultArgumentName": 1,
+             |    "$DefaultArgumentName": 2
+             |  }
+             |}
+           """.stripMargin.parseJson
```

Comment:
For such simple json stricture (here and in other places above), maybe it would be better to use Json AST directly. E.g.

```scala
JsObject(NestedParentArgumentName → JsObject(
  NonDefaultArgumentName → JsNumber(1),
  DefaultArgumentName → JsNumber(2)))
```

WDYT?